File: algorithms/sort/mergeSort.cpp
/*
** mergeSort.cpp
** 2020-02-26
*/

#include <stdio.h>

int sorted[10];

void merge(int arr[], int start, int mid, int end) {
	
	int i = start;
	int j = mid+1;
	int k = start;
	
	while(i <= mid && j <= end) {
		if(arr[i] < arr[j]) {
			sorted[k] = arr[i];
			i++;
		} else {
			sorted[k] = arr[j];
			j++;
		}
		k++;
	}
	
	if(i > mid) {
		for(int t=j; t<=end; t++) {
			sorted[k] = arr[t];
			k++;			
		}
	} else {
		for(int t=i; t<=mid; t++) {
			sorted[k] = arr[t];
			k++;			
		}
	}
	
	for(int t=start; t<=end; t++) {
		arr[t] = sorted[t];
	}
}

void mergeSort(int arr[], int start, int end) {
	if(start < end) {
		mergeSort(arr, start, (start+end)/2);
		mergeSort(arr, (start+end)/2+1, end);
		merge(arr, start, (start+end)/2, end);		
	}
}

int main(void) {
	int a[10] = {6, 4, 8, 2, 1, 10, 7, 9, 3};
	mergeSort(a, 0, 8);
	
	for(int i=0; i<9; i++) {
		printf("%d ", a[i]);
	}
	return 0;
}

File: BOJ/src/algorithms/MergeSort.java
package algorithms;

import java.util.Arrays;

public class MergeSort {

	static int[] sorted;
	
	public static void main(String[] args) {
		int[] list = {1,5,3,2,8,4,9,6,7,6,4};
		
		sorted = new int[list.length];
		mergeSort(list, 0, list.length-1);
		
		System.out.println(Arrays.toString(sorted));
	}

	private static void mergeSort(int[] list, int left, int right) {
		int mid = (left+right)/2;
		
		if(left < right) {
			mergeSort(list, left, mid);
			mergeSort(list, mid+1, right);
			merge(list, left, mid, right);
		}
	}

	private static void merge(int[] list, int left, int mid, int right) {
		int i = left;
		int j = mid+1;
		int k = left;
		
		while(i<=mid && j<=right) {
			if(list[i] <= list[j]) {
				sorted[k++] = list[i++];
			} else {
				sorted[k++] = list[j++];
			}
		}
		  
		if(i>mid) {
			for(int idx=j; idx<=right; ++idx) { sorted[k++] = list[idx]; }
		} else {
			for(int idx=i; idx<=mid; ++idx) { sorted[k++] = list[idx]; }
		}				
		
		for(int idx=left; idx<=right; ++idx) {
			list[idx] = sorted[idx];
		}
	}	
}